pkg/quarterdeck/db/models/projects.go

```diff
@@ -0,0 +1,269 @@
+)
+
+// Save an organization project mapping to the database by creating a record.
+// Organization project mappings can only be created and deleted, not updated, so if the
+// mapping already exists an error is returned.
+//
+// NOTE: because this is a security condition, the OrgID in the OrganizationProject
+// model must come from the user claims and not from user input!
+func (op *OrganizationProject) Save(ctx context.Context) (err error) {
+	switch {
+	case ulids.IsZero(op.OrgID):
+		return ErrMissingOrgID
+	case ulids.IsZero(op.ProjectID):
+		return ErrMissingProjectID
+	}
+
+	var tx *sql.Tx
+	if tx, err = db.BeginTx(ctx, nil); err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	now := time.Now()
+	op.SetCreated(now)
+	op.SetModified(now)
+
+	params := make([]any, 4)
+	params[0] = sql.Named("orgID", op.OrgID)
+	params[1] = sql.Named("projectID", op.ProjectID)
+	params[2] = sql.Named("created", op.Created)
+	params[3] = sql.Named("modified", op.Modified)
+
+	if _, err = tx.Exec(insertOrgProjSQL, params...); err != nil {
+		var dberr sqlite3.Error
+		if errors.As(err, &dberr) {
+			if dberr.Code == sqlite3.ErrConstraint {
+				return constraint(dberr)
+			}
+		}
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// Exists checks if an organization project mapping exists in order to verify that a
+// project is allowed to be associated with an APIKey or other claims resource for the
+// user with the specified OrgID claims. Only the OrgID and ProjectID are used for this
+// so no preliminary fetch from the database is required to execute the query.
+func (op *OrganizationProject) Exists(ctx context.Context) (ok bool, err error) {
+	var tx *sql.Tx
+	if tx, err = db.BeginTx(ctx, &sql.TxOptions{ReadOnly: true}); err != nil {
+		return false, err
+	}
+	defer tx.Rollback()
+
+	if ok, err = op.exists(tx); err != nil {
+		return ok, err
+	}
+	tx.Commit()
+	return ok, err
+}
+
+const (
+	orgProjExistsSQL = "SELECT EXISTS(SELECT 1 FROM organization_projects WHERE organization_id=:orgID AND project_id=:projectID);"
+)
+
+// Check if a project is associated with an organization inside of a transaction.
+func (op *OrganizationProject) exists(tx *sql.Tx) (ok bool, err error) {
+	if err = tx.QueryRow(orgProjExistsSQL, sql.Named("orgID", op.OrgID), sql.Named("projectID", op.ProjectID)).Scan(&ok); err != nil {
+		return false, err
+	}
+	return ok, nil
+}
+
+const listProjectsSQL = `WITH projects AS (
+	SELECT op.organization_id, op.project_id, op.created, op.modified, count(k.id) as apikeys_count
+		FROM organization_projects op
+		LEFT JOIN api_keys k ON op.organization_id=k.organization_id AND op.project_id = k.project_id
+	GROUP BY op.organization_id, op.project_id),
+project_counts AS (
+	SELECT p.*, count(r.id) as revoked_count
+		FROM projects p
+		LEFT JOIN revoked_api_keys r ON p.organization_id=r.organization_id AND p.project_id=r.project_id
+	GROUP BY p.organization_id, p.project_id)
+	SELECT * FROM project_counts`
```

Comment:
The line breaks and indents definitely help with readability 👍

Reply:
It's a pretty hefty query; that's for sure.